anvil-python/anvil/utils.py

```diff
@@ -43,3 +45,20 @@ def __call__(self, *args, **kwargs):  # type: ignore[no-untyped-def]
             LOG.warn(message)
             LOG.error("Error: %s", e)
             sys.exit(1)
+
+
+def machines_missing_juju_controllers() -> list[str]:
+    result = subprocess.run(
+        ["juju", "show-controller", "anvil-controller", "--format", "json"],
+        capture_output=True,
+    )
+    controllers = json.loads(result.stdout)
+    controller_machines = set(
+        controllers["anvil-controller"]["controller-machines"].keys()
+    )
+
+    machines_res = subprocess.run(
+        ["juju", "machines", "--format", "json"], capture_output=True
+    )
+    machines = set(json.loads(machines_res.stdout)["machines"].keys())
+    return list(machines - controller_machines)
```

Comment:
I personally prefer using methods over operators (`list(machines.difference(controller_machines))`), but it doesn't really change anything here.

Reply:
Fine with me! Changed.